WIP: aste Replay Mode tutorial on elastic flap case

richahert opened the pull request:
This PR adds an tutorial for aste. From the OpenFOAM-FEniCS-FSI tutorial OpenFOAM is replaced with aste that sends previously exported results to FEniCS.

richahert committed:
commit so i can pull the master from upstream

richahert committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	FSI/cylinderFlap/OpenFOAM-FEniCS/Solid/cyl-flap.py

richahert committed:
add readme

richahert committed:
create first Readme

richahert committed:
Merge remote-tracking branch 'upstream/master' into aste

richahert committed:
precice config for aste fsi

richahert committed:
precice config for aste fsi

richahert committed:
precice config for aste fsi

richahert committed:
new precice config

richahert committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	FSI/flap_perp/OpenFOAM-CalculiX/precice-config_serial.xml
#	FSI/flap_perp/OpenFOAM-FEniCS/precice-config_serial.xml

richahert committed:
Merge branch 'master' into aste

# Conflicts:
#	FSI/flap_perp/OpenFOAM-FEniCS/precice-config.xml

richahert committed:
Add basic files for aste tutorial

richahert committed:
add gitignore, rename forced to data

richahert committed:
adds in gitignore

richahert committed:
Add instructions. Extended way not finished yet.

richahert committed:
add links and missing infos

richahert committed:
Update README.md

richahert committed:
Merge remote-tracking branch 'upstream/develop' into aste

uekerman commented:
Try to merge the develop branch into your branch https://github.com/richahert/tutorials/tree/aste. Then, the changes here should be easier to overlook.
What is your overall plan, to create a new tutorial ASTE-OpenFOAM or to explain how to generate such a tutorial from OpenFOAM-FEniCS?

BenjaminRodenberg reviewed (changes requested):
`FSI/flap_perp/ASTE/Solid/perp-flap.py` is a copy from `FSI/cylinderFlap/OpenFOAM-FEniCS/Solid/perp-flap.py`. We ask the user explicitly to copy this file in our `README.md`, therefore, we could avoid this duplicate (related #28). I also think it is a bit counter-intuitive to ask the user to copy the file even though it is already provided in the folder.

For `precice-config.xml` the situation is different: Changes are required and therefore, we should provide this file. The user may directly use it (quick path) or try to come up with his or her own `ASTE/precice-config.xml` derived from `OpenFOAM-FEniCS/precice-config.xml`.

The remaining changes are minor formatting / cleanup changes:

* A comment on markdown formatting: Note that one backtick is enough for inline code, if you want to have code blocks, use three backticks (see [here](https://guides.github.com/features/mastering-markdown/)).
* Don't use `from fenics import *`. Better use `from fenics import ...`. This will require you to add `XDMFFile`, `TensorFunctionSpace` and `LocalSolver` to the `from fenics import ...` in `perp-flap.py`.

**Note:**

I'm currently stuck with a segmentation fault, when I try to run `preciceMap -v -c precice-config.xml -p A --mesh preCICE-output/Solid-fenics --vectordata`. It looks like an error that is not being treated properly within preCICE, but it might also be wrong usage on my side. Therefore, I will submit this review now and do another review as soon as I'm able to run the case.

richahert committed:
Merge remote-tracking branch 'upstream/develop' into aste

richahert committed:
revert changes

richahert committed:
delete strange file

richahert committed:
remove file that is not relevant

richahert committed:
Update FSI/flap_perp/ASTE/README.md

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert commented on FSI/flap_perp/ASTE/README.md:
````diff
@@ -0,0 +1,95 @@
+- OpenFOAM, e.g. [OpenFOAM 7](https://openfoam.org/version/7/)
+- [OpenFOAM Adapter](https://github.com/precice/openfoam-adapter/wiki/Building) matching the OpenFOAM version.
+
+Make sure to add aste/build to the ```PATH``` such that the python scripts and ```preciceMap``` can be found from anywhere on your system.
````
That was suggested for the Bunny-Bloodcell tutorial so I thought it's most straightforward to do it the same way. I agree though that it is not a very nice and clean way.

richahert committed:
Update FSI/flap_perp/ASTE/README.md formatting

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert committed:
Apply suggestions from code review

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert committed:
delete unnecessary comments

richahert committed:
revert changes to upstream develop

richahert committed:
revert to upstream/develop

richahert committed:
commit suggestion

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert committed:
PEP8 formatting

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert committed:
revert to upstream

richahert committed:
remove profiling

richahert committed:
formatting

richahert commented on FSI/flap_perp/OpenFOAM-FEniCS/Solid/perp-flap.py:
```diff
@@ -97,8 +107,8 @@ def Neumann_Boundary(x, on_boundary):
 
-alpha_m = Constant(0.2)
-alpha_f = Constant(0.4)
+alpha_m = Constant(0.0)
+alpha_f = Constant(0.0)
```
I think this should be changed since the alpha method showed bad stability with 0.2 and 0.4 in the FSI3 case.

BenjaminRueth replied:
Sounds like a good reason to change it. 

Note: With Waveform Relaxation, we might be able to go back to the alpha method with 0.2 and 0.4. But that's still a quite long way to go (see https://github.com/precice/precice/issues/133)

richahert committed:
remove export

Co-Authored-By: Benjamin Rüth <benjamin.rueth@tum.de>

richahert committed:
add solvername and some formatting

richahert committed:
-f flag was chaged to --tag or -t

BenjaminRodenberg committed:
avoid from fenics import *

BenjaminRodenberg committed:
Wording and formatting

BenjaminRodenberg committed:
Remove perp-flap.py

Should be copied from the OpenFOAM-FEniCS tutorial to this location accordinato the the README.md

BenjaminRodenberg committed:
Merge branch 'aste' of github.com:richahert/tutorials into richahert_aste

BenjaminRodenberg reviewed (changes requested):
- [x] currently aste seems to not provide the option `--vectordata`. I am using https://github.com/precice/aste/pull/9/commits/84ebcbcb8903aea17495a487ed0309cd43aacbd8. Therefore, I cannot run the replay mode.
- [ ] why is `FSI/cylinderFlap/OpenFOAM-FEniCS/Postprocessing/plot_tip.py` deleted? Is it not needed? In any case: I am not sure whether removing this file should be done within this PR.

:heavy_check_mark: the instructions in the `README.md` are helpful and clear

Note @richahert: I can take care of applying the changes on my own. I just need some help with the `--vectordata` option.

BenjaminRueth commented:
> Note @richahert: I can take care of applying the changes on my own. I just need some help with the `--vectordata` option.

The option seems to have been removed in https://github.com/richahert/aste/commit/94537d4bff9c0f389db37e70c8a8bfeabd4ec255. Whether we are dealing with scalar or vector data is detected automatically.

If I remove the `--vectordata` option, I run into a Segmentation Fault:

```
benjamin@benjamin-ThinkPad-X1-Yoga-2nd:~/tutorials/FSI/flap_perp/ASTE$ preciceMap -v -c precice-config.xml -p A --mesh preCICE-output/Solid-fenics
2020-05-08 15:35:47,693 VERBOSE-1 [default] Basename: preCICE-output/Solid-fenics
2020-05-08 15:35:47,693 DEBUG [default] [benjamin@unknown-host] [std::vector<std::__cxx11::basic_string<char> > getMeshes(std::__cxx11::string, int)] [/home/benjamin/aste/src/preciceMap.cpp:48] Meshes:
(0) 15:35:47 [impl::SolverInterfaceImpl]:120 in configure: This is preCICE version 2.0.2
(0) 15:35:47 [impl::SolverInterfaceImpl]:121 in configure: Revision info: v2.0.1-146-g15fb6857-dirty
(0) 15:35:47 [impl::SolverInterfaceImpl]:122 in configure: Configuring preCICE with configuration "precice-config.xml"
(0) 15:35:47 [impl::SolverInterfaceImpl]:123 in configure: I am participant "A"
[benjamin-ThinkPad-X1-Yoga-2nd:05121] *** Process received signal ***
[benjamin-ThinkPad-X1-Yoga-2nd:05121] Signal: Segmentation fault (11)
[benjamin-ThinkPad-X1-Yoga-2nd:05121] Signal code: Address not mapped (1)
[benjamin-ThinkPad-X1-Yoga-2nd:05121] Failing at address: (nil)
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f0d25207890]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 1] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE5c_strEv+0x0)[0x7f0d24a86470]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 2] preciceMap(_ZN2el4base14MessageBuilderlsERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x1c)[0x55bfb106f1bc]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 3] preciceMap(_ZN2el4base6WriterlsINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEERS1_RKT_+0x33)[0x55bfb106ffab]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 4] preciceMap(_Z12readMainFileR4MeshRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEb+0x126)[0x55bfb10641b9]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 5] preciceMap(_Z8readMeshRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEb+0x60)[0x55bfb1064b30]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 6] preciceMap(main+0x3dd)[0x55bfb10650b4]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f0d24378b97]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] [ 8] preciceMap(_start+0x2a)[0x55bfb106364a]
[benjamin-ThinkPad-X1-Yoga-2nd:05121] *** End of error message ***
Segmentation fault (core dumped)
```

I think this is an error on my side, since the data is not provided correctly. I will try to debug this now and I might comment in https://github.com/precice/aste/pull/9 or open an issue in https://github.com/precice/aste. A Segmentation Fault should, of course, be avoided.

BenjaminRodenberg committed:
Clearify modifications of precice-config.xml

BenjaminRodenberg committed:
Remove unneeded --vectordata option

BenjaminRueth commented:
I forgot converting the data using `precice_to_aste.py`. After doing this and removing the `--vectordata` option, everything worked perfectly :+1: 

I will report the issue with the segmentation upstream. This is not related to this PR.

The only remaining point is whether or not (and where) `FSI/cylinderFlap/OpenFOAM-FEniCS/Postprocessing/plot_tip.py` should be deleted. Then this PR is ready to merge. 

I would suggest to wait for https://github.com/precice/aste/pull/9 before merging this PR, since without the feature being implemented in aste this tutorial does not really help.

MakisH commented:
This now would need to be adjusted to the [new tutorials structure](https://precice.org/community-contribute-to-precice.html#contributing-tutorials).

BenjaminRodenberg committed:
Merge branch 'develop' into aste

BenjaminRodenberg committed:
Migrating FSI/flap_perp/ASTE to the new tutorials structure

* Moves files that are not conflicting from FSI/flap_perp/ASTE to perpendicular-flap
* perpendicular-flap/fluid-aste/README.md describes usage of aste for replay mode. This information needs to be incorporated into the new structure or moved to preCICE.org.
* FSI/flap_perp/ASTE/Solid/precice-adapter-config-fsi-s.json is moved to perpendicular-flap/solid-fenic. Here integration is still a TODO.
* Overwrites perpendicular-flap/solid-fenics/solid.py with FSI/flap_perp/OpenFOAM-FEniCS/Solid/perp-flap.py. Here integration into new structure is still a TODO.
* Overwrites perpendicular-flap/precice-confix.xml with FSI/flap_perp/OpenFOAM-FEniCS/precice-config.xml. Here integration into new structure is still a TODO.

BenjaminRodenberg commented on perpendicular-flap/fluid-aste/README.md:
```diff
@@ -0,0 +1,94 @@
+
+#### The quick way to run 
+
+1. Copy `perp_flap.py` from `tutorials/FSI/flap_perp/OpenFOAM-FEniCS/Solid` into `tutorials/FSI/flap_perp/ASTE/Solid`.
```
Copying files should not be necessary anymore with the new structure.

BenjaminRodenberg commented on perpendicular-flap/fluid-aste/README.md:
```diff
@@ -0,0 +1,94 @@
+
+1. Copy `perp_flap.py` from `tutorials/FSI/flap_perp/OpenFOAM-FEniCS/Solid` into `tutorials/FSI/flap_perp/ASTE/Solid`.
+
+2. Run `python3 Solid/perp-flap.py` and `preciceMap -v -c precice-config.xml -p A --mesh preCICE-output/Solid-fenics` in two terminals.
```
`perp-flap.py` is now called `solid.py`

BenjaminRodenberg replied:
Renamed this.

BenjaminRodenberg commented on perpendicular-flap/fluid-aste/README.md:
```diff
@@ -0,0 +1,94 @@
+
+- Export the output at the Fluid participant. Instead of `preCICE-output/Solid-fenics.*` the files containing the forces will be `preCICE-output/Fluid-Mesh-Faces-Fluid.*`
+
+- Export the forces in the [Calculix-Openfoam-tutorial](https://github.com/precice/tutorials/tree/master/FSI/flap_perp/OpenFOAM-CalculiX) with the same geometry and feed them to FEniCS with aste. 
```
This link has to be updated.

BenjaminRodenberg replied:
Done.

BenjaminRodenberg commented on perpendicular-flap/precice-config-aste.xml:
```diff
@@ -0,0 +1,52 @@
+      <export:vtk every-n-time-windows="1" directory="vtkA/" normals="0"/>
+    </participant>
+    
+    <participant name="fenics">
```
Example for "old" terminology: This should be named `"solid"`.

BenjaminRodenberg replied:
Also updated.

BenjaminRodenberg commented on perpendicular-flap/solid-fenics/precice-adapter-config-fsi-s-aste.json:
```diff
@@ -0,0 +1,8 @@
+  "config_file_name": "../precice-config.xml",
+  "interface": {
+      "coupling_mesh_name": "Solid",
+      "read_data_name": "Data"
```
In replay mode we only read data from ASTE with the solid solver. This is an important difference compared to the normal tutorial.

BenjaminRodenberg reviewed (changes requested):
Todos I identified when moving to the new tutorials structure:

- [x] We might be able to remove `Allclean` and `.gitignore`. I'm not sure whether there is anything specific to the replay-mode.
- [ ] `fluid-aste/README.md` explains how the replay mode is used. This information could me merged into `perpendicular-flap/README.md` or we could provide a full explanation of the replay more on the website and only provide specific information here in this tutorial.
- [ ] `precice-config-aste.xml` has to be updated to the new naming. Ideally we could get rid of it and just use `precice-config.xml` for replay-mode as well.

BenjaminRodenberg commented:
I moved the files from the old structure to the new structure. This PR is currently not working. It might be helpful to go back to 
a7a3dd5 to see what the expected behavior is and to find out how paths, variable names etc. have to be mapped to the new structure.

BenjaminRodenberg committed:
Update obvious changes. Whitespaces etc.

BenjaminRodenberg committed:
Some more obvious changes.

BenjaminRodenberg committed:
Update configs to new naming scheme.

BenjaminRodenberg committed:
Update solver.

BenjaminRodenberg committed:
Clean up the diff.

BenjaminRodenberg committed:
Partially update to new tutorials structure

BenjaminRodenberg committed:
More cleaning up of the diff.

BenjaminRodenberg committed:
Fix md file formatting.

BenjaminRodenberg committed:
Fix ShellCheck.

BenjaminRodenberg committed:
Fix dead links.

BenjaminRodenberg committed:
Fix format.

BenjaminRodenberg committed:
Remove unnecessary clean script.

BenjaminRodenberg committed:
Fix link.

BenjaminRodenberg commented on perpendicular-flap/fluid-aste/README.md:
```diff
@@ -0,0 +1,96 @@
+Copy the `preCICE-output` folder to the root directory of the aste tutorial `tutorials/perpendicular-flap/fluid-aste`.
+To convert the files to the correct format, open the `preCICE-output` folder and run
+
+`precice_to_aste.py Solid-fenics -n 500 -t Forces0 --datadim 3`
```
I'm currently getting stuck here. Where is the script `precice_to_aste.py`? I cannot find it in https://github.com/precice/aste/tree/develop. @fsimonis any idea?

DavidSCN replied:
@BenjaminRodenberg https://github.com/precice/aste/pull/9/files

BenjaminRodenberg commented:
Since this PR is idle some time, here is a quick status update:

* Currently ASTE is under a major revision, therefore this PR is waiting. Replay mode is still on the list of features that we would like to implement in the future.
* We might implement replay mode by modifying this PR or by creating a new one (using a big part of this PR as a starting point). We should make sure that the contributions by @richahert also show up in the history of a new PR, if we crate one.
* In the end, replay mode should work with any `solid-*` solver. Therefore, the only change that should be applied to the FEniCS solver, for example, is that a different configuration file should be used (not `precice-config.xml` but `precice-config-aste.xml`). I guess that the (many) code changes in `solid.py` (computation of stresses, removing checkpointing) were due to development. The adapter should be able to handle the correct treatment of checkpointing on its own. Computation of stresses is not required, if I understand it correctly.
* @DavidSCN and @kursatyurt will take care of replay mode. I'm available for reviews, if needed.

MakisH commented:
We now have documentation on this: https://precice.org/tooling-aste.html#replay-mode This PR is no longer needed.
